### PR TITLE
[KLC-1964] Implement complete rollback logic to revert account balances and history

### PR DESCRIPTION
**Pull request by Beroni**
## Summary
Refactor blockchain rollback logic to include account balance and account history reversion, implementing a complete 4-step rollback process triggered when blocks are removed from Elasticsearch due to consensus failures.

## Problem
When a rollback occurred due to network consensus failure, only transactions and blocks were removed from Elasticsearch indices. Account balances and account history entries were not reverted, causing balance inconsistencies that persisted after the rollback. This issue was reported via Slack thread where a transaction executed during a consensus failure period (blocks 26364634-26364635) left incorrect account balances.

## Solution
Implemented a comprehensive 4-step rollback process in the Elasticsearch indexer:

1. **Revert Account Balances** - Query account history by timestamp to find all affected accounts, then revert each account to its previous state
2. **Remove Account History** - Delete account history entries for the rolled-back block using timestamp-based bulk removal
3. **Remove Transactions** - Clean up transaction records from the block (existing functionality)
4. **Remove Block Header** - Remove the block header from indices (existing functionality)

### Optimization Strategy
The implementation uses an optimized query approach:
- Single timestamp-based search query to retrieve all account history entries from the rolled-back block
- For each affected account, fetch the last 2 history entries (sorted DESC by timestamp)
- Determine correct reversion based on 3 cases:
  - **No history found**: Log warning (edge case)
  - **Single history entry**: Account was created in this block → revert balance to 0
  - **Multiple entries**: Account existed before → revert to penultimate (second-to-last) state

This approach minimizes Elasticsearch queries and leverages indexed data rather than computing state from scratch.

**New Methods & Functionality:**
- `indexer/elasticClient.go` (+82 lines):
  - `DoBulkRemoveByTimestamp()` - Bulk delete documents by timestamp query
  - `DoSearch()` - Execute search queries with custom parameters
  - `DoUpdate()` - Update document fields by ID

- `indexer/elasticProcessor.go` (+199 lines):
  - `RemoveAccountsHistory(blockTimestamp)` - Delete all account history entries for a specific timestamp
  - `RevertAccountBalances(blockTimestamp)` - Revert account balances to previous state
  - `extractAccountHistoryEntries()` - Parse search results into structured account history data
  - `revertSingleAccountBalance()` - Handle individual account reversion with 3-case logic

- `indexer/queries.go` (+10 lines):
  - `prepareTimestampForBulkRemove()` - Build Elasticsearch term query for timestamp-based deletion

- `indexer/data/account.go` (+7 lines):
  - Added `BlockNum` field to `AccountBalanceHistory` struct for tracking

**Enhanced Orchestration:**
- `indexer/workItems/workItemRemoveBlock.go` (+36 lines):
  - Refactored to implement 4-step rollback workflow
  - Added comprehensive logging with nonce and timestamp tracking for each step
  - Proper error handling and step sequencing

**Interface Updates:**
- `indexer/interface.go`: Added `RemoveAccountsHistory()` and `RevertAccountBalances()` method signatures
- `indexer/workItems/interface.go`: Updated `removeIndexer` interface

**Mock Implementations:**
- `indexer/mock/databaseWriterStub.go` (+15 lines): Added stubs for new DB client methods
- `indexer/mock/elasticProcessorStub.go` (+10 lines): Added stubs for rollback operations

## Testing
- [x] All existing tests pass
- [x] Manual testing performed:
  - Verified 4-step process executes in correct order
  - Confirmed account balances revert to previous state
  - Tested edge case: new account created in rolled-back block correctly reverts to balance 0
  - Validated account history entries are properly removed by timestamp
  - Confirmed logs show nonce and timestamp for each rollback step

## Breaking Changes
None

## Related Issues
Fixes [KLC-1964](https://klever-io.atlassian.net/jira/software/c/projects/KLC/boards/78?assignee=61eec400571f7600714f0913&selectedIssue=KLC-1964)

## Checklist
- [x] Code follows project style guidelines
- [x] Self-reviewed code for logic and edge cases
- [x] Tests added/updated and passing
- [x] Added comprehensive code comments explaining rollback logic
- [x] Commit message follows Conventional Commits format (`chore: refactor rollback logic`)
- [x] No unnecessary files included
- [x] Breaking changes documented (None)
- [x] Configuration dependencies documented
- [x] Performance optimizations implemented (timestamp-based queries, bulk operations)

[KLC-1964]: https://klever-io.atlassian.net/browse/KLC-1964?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: indexer/elasticProcessor.go [hunk 1]
```diff
@@ -354,12 +355,208 @@ func (ei *elasticProcessor) RemoveTransactions(blk nodeData.HeaderHandler) error
...
+				"must": []templates.Object{
+					{
+						"term": templates.Object{
+							"address": address,
+						},
+					},
+					{
+						"range": templates.Object{
+							"timestamp": templates.Object{
+								"lte": currentBlockTimestamp,
```
[the remark this answers is not in the document]
Comment (Beroni) [reply indexer/elasticProcessor.go]: no need, this already tested.

File: indexer/data/account.go [hunk 2]
```diff
@@ -166,3 +166,10 @@ type ValidatorInfo struct {
 	Logo           string `json:"logo"`
 	URIs           []*URI `json:"uris"`
 }
+
+type AccountHistoryEntry struct {
+	Address       string
+	Balance       int64
+	FrozenBalance int64
+	Timestamp     int64
```
[the remark this answers is not in the document]
Comment (Beroni) [reply indexer/data/account.go]: no need json tags for this struct.

File: indexer/workItems/workItemRemoveBlock.go [hunk 3]
```diff
@@ -28,21 +28,37 @@ func (wirb *itemRemoveBlock) IsInterfaceNil() bool {
...
+		err = wirb.indexer.RemoveTransactions(blk)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Step 4: Remove block header
+	err = wirb.indexer.RemoveHeader(wirb.headerHandler)
 	if err != nil {
-		log.Warn("itemRemoveBlock.Save could not remove block transactions", "error", err.Error())
```
Comment (fbsobreira) [inline indexer/workItems/workItemRemoveBlock.go:45]: why removed the wraning?

File: indexer/elasticProcessor.go [hunk 4]
```diff
@@ -1538,6 +1710,7 @@ func (ei *elasticProcessor) buildAccountInfo(userAccount *data.Account, blockTim
 		Allowance:       ei.getAllowanceWithPendingRewards(userAccount.UserAccount),
 		Permissions:     permissions,
 		Timestamp:       time.Duration(blockTimestamp * 1000),
+		UpdatedAt:       time.Duration(blockTimestamp * 1000),
```
Comment (fbsobreira) [inline indexer/elasticProcessor.go]: if `updatedAt` is the same of `timestamp` makes no sense to add it.

File: indexer/elasticProcessor.go [hunk 5]
```diff
@@ -354,12 +357,181 @@ func (ei *elasticProcessor) RemoveTransactions(blk nodeData.HeaderHandler) error
...
+	}
+
+	return accounts, nil
+}
+
+// revertSingleAccountBalanceFromTrie reads the current balance from the account trie
+// and updates the accounts index with that balance.
+// This assumes RevertStateToBlock has already been called, so the trie is at the correct state.
+// oldBalance and oldFrozenBalance are the values from the accounts-history index (for logging comparison).
+func (ei *elasticProcessor) revertSingleAccountBalanceFromTrie(address string, oldBalance, oldFrozenBalance int64) error {
```
Comment (fbsobreira) [inline indexer/elasticProcessor.go:471]: Im not confident with this solution, it only care for KLV balance, what about a KDA transfer, or freeze/unfreeze? Im not sure if is possible but best way would be indexing as we do the indexing saving that was changed and reload that info. or have another mechanism...

**Review by fbsobreira — CHANGES_REQUESTED** [review]

File: indexer/elasticProcessor.go (hunk 4 above)
Comment (Beroni) [reply indexer/elasticProcessor.go]: updatedAt is different from timestamp.
The timestamp value is not updated when `serializedDataForUpdateAccounts` is called, because timestamp is treated as the createdAt value.

File: indexer/workItems/workItemRemoveBlock.go (hunk 3 above)
Comment (Beroni) [reply indexer/workItems/workItemRemoveBlock.go:45]: I removed those logs thinking they were the ones I had added to test the revert. I'm adding them back now.

File: indexer/elasticProcessor.go (hunk 5 above)
Comment (Beroni) [reply indexer/elasticProcessor.go:471]: The motivation behind this task was specifically the KLV balances, which directly impact only the account and account-history indexes — as highlighted in the report attached to the task.
In the future, we can expand the revert mechanism to include  accountskda index

Comment (Beroni) [reply indexer/elasticProcessor.go:471]: During my research, I also checked whether Elasticsearch had any native mechanism that could help us. According to the official documentation, Elasticsearch does not support rollbacks. It does support snapshots, but taking a snapshot for every block wouldn’t be feasible.

Since this bug is very unlikely to happen—especially after the optimization updates we’ve implemented—we can put this task on hold and look for a better solution.

File: indexer/elasticProcessor.go [hunk 6]
```diff
@@ -354,12 +357,181 @@ func (ei *elasticProcessor) RemoveTransactions(blk nodeData.HeaderHandler) error
...
+	if len(accounts) == 0 {
+		log.Debug("RevertAccountBalances: no accounts to revert", "blockTimestamp", blockTimestamp)
+		return nil
+	}
+
+	// For each account modified in this block, get current balance from trie and update index
+	for _, accountInfo := range accounts {
+		err := ei.revertSingleAccountBalanceFromTrie(accountInfo.Address, accountInfo.Balance, accountInfo.FrozenBalance)
+		if err != nil {
+			log.Debug("RevertAccountBalances: failed to revert account",
```
Comment (fbsobreira) [inline indexer/elasticProcessor.go]: As we don't propagate the error we should log as warn

File: indexer/elasticProcessor.go [hunk 7]
```diff
@@ -31,6 +32,8 @@ import (
 )
 
 const numDecimalsInFloatBalance = 10 //?
+const revertAccountBatchSize = 10_000
```
Comment (fbsobreira) [inline indexer/elasticProcessor.go]: lets add a comment on the 10k choice (ES query size limit... ), also lets check the line above with "//??

File: indexer/elasticProcessor.go [hunk 8]
```diff
@@ -354,12 +357,181 @@ func (ei *elasticProcessor) RemoveTransactions(blk nodeData.HeaderHandler) error
...
 	return ei.elasticClient.DoBulkRemove(txIndex, encodedTXsHashes)
 }
 
+// RemoveAccountsHistory will remove account history entries for a specific timestamp
+func (ei *elasticProcessor) RemoveAccountsHistory(blockTimestamp int64) error {
+	if !ei.isIndexEnabled(accountsHistoryIndex) {
+		return nil
+	}
+
+	return ei.elasticClient.DoBulkRemoveByTimestamp(accountsHistoryIndex, blockTimestamp*1000)
```
Comment (fbsobreira) [inline indexer/elasticProcessor.go]: `blockTimestamp*1000` multiplied been used as param and in `RevertAccountBalances` L#384 the multiplication its been done in query.
Lets make this consistent. Also, consider using a `toMilliseconds(timestamp int64) time.Duration`

**Review by fbsobreira — CHANGES_REQUESTED** [review]

File: indexer/elasticProcessor.go (hunk 6 above)
Comment (Beroni) [reply indexer/elasticProcessor.go]: ok

File: indexer/elasticProcessor.go (hunk 8 above)
Comment (Beroni) [reply indexer/elasticProcessor.go]: ok

File: indexer/elasticProcessor.go (hunk 7 above)
Comment (Beroni) [reply indexer/elasticProcessor.go]: ok

**Review by fbsobreira — APPROVED** [review]

**Review by RomuloSiebra — APPROVED** [review]
